src/node/interfaces.cpp

```diff
@@ -971,24 +971,32 @@ class MinerImpl : public Mining
         return BlockRef{tip->GetBlockHash(), tip->nHeight};
     }
 
-    BlockRef waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
+    std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
     {
         if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
+        std::optional<uint256> tip_hash;
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
             notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                 // We need to wait for m_tip_block to be set AND for the value
                 // to differ from the current_tip value.
-                return (notifications().TipBlock() && notifications().TipBlock() != current_tip) || chainman().m_interrupt;
+                tip_hash = notifications().TipBlock();
+                return (tip_hash && tip_hash != current_tip) || chainman().m_interrupt;
             });
         }
+
+        if (!tip_hash || chainman().m_interrupt) return {};
+
         // Must release m_tip_block_mutex before locking cs_main, to avoid deadlocks.
         LOCK(::cs_main);
-        return BlockRef{chainman().ActiveChain().Tip()->GetBlockHash(), chainman().ActiveChain().Tip()->nHeight};
+        return BlockRef{*Assume(tip_hash), Assume(chainman().ActiveChain().Tip())->nHeight};
     }
 
     std::unique_ptr<BlockTemplate> createNewBlock(const BlockCreateOptions& options) override
     {
+        // Ensure m_tip_block is set so consumers of BlockTemplate can rely on that.
+        if (!waitTipChanged(uint256::ZERO, MillisecondsDouble::max())) return {};
```

Comment:
The default for the timeout argument is `MillisecondsDouble::max()`, so:

```suggestion
        if (!waitTipChanged(uint256::ZERO)) return {};
```

Reply:
It seems the compiler doesn't know about this default inside the implementation file.

Reply:
TIL https://flylib.com/books/en/2.887.1.53/1/

Possible solutions I see:
* drop the default altogether and have all callers use 2 arguments
* use the "non-virtual interface idiom" as described in the chapter above (it is from the book "Effective C++" by Scott Meyers)
* use overload instead of default arguments. Have two methods `waitTipChanged(hash)` and `waitTipChanged(hash, timeout)` (the first could call the second internally).

Reply:
I don't think we should drop the default, because `MillisecondsDouble::max()` is tedious. It seems fine that the interface does it internally, which is simpler than the other two solutions.

Reply:
I could also do this:

```diff
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index 66ce64b1c3..b0735d497c 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -971,7 +971,7 @@ public:
         return BlockRef{tip->GetBlockHash(), tip->nHeight};
     }
 
-    std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
+    std::optional<BlockRef> waitTipChanged(uint256 current_tip, MillisecondsDouble timeout = MillisecondsDouble::max()) override
     {
         if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
         std::optional<uint256> tip_hash;
@@ -1005,7 +1005,7 @@ public:
     std::unique_ptr<BlockTemplate> createNewBlock(const BlockCreateOptions& options) override
     {
         // Ensure m_tip_block is set so consumers of BlockTemplate can rely on that.
-        if (!waitTipChanged(uint256::ZERO, MillisecondsDouble::max())) return {};
+        if (!waitTipChanged(uint256::ZERO)) return {};
 
         BlockAssembler::Options assemble_options{options};
         ApplyArgsManOptions(*Assert(m_node.args), assemble_options);
```

But that doesn't seem nicer.

Reply:
My takeaway from the article is that having a default for the base class is fine, and that it's better to not use a default for the inherited classes (but it could get ignored).

The patch in my previous comment would be bad. But the current code that calls `waitTipChanged(uint256::ZERO, MillisecondsDouble::max())` should be fine.

Reply:
In commit "Have createNewBlock() wait for a tip" (854766b886b6c2ee564b867db64bdbce80064b3b)

Another alternative but probably not better is:

```diff
-        if (!waitTipChanged(uint256::ZERO, MillisecondsDouble::max())) return {};
+        if (!static_cast<Mining*>(this)->waitTipChanged(uint256::ZERO)) return {};
```

I think in longer term after we add an initial waitNext() method and make improvements to it, it will be natural to move more functionality out of Mining class methods into node/miner.cpp or somewhere like that and these methods will get simpler. The node/interfaces.cpp file is mostly meant to contain glue code and wrap other functions, not implement real functionality itself. Since a lot of code needs to wait, I imagine there will be some new waiting utility function this could call instead of waitTipChanged, but using waitTipChanged for now here should be fine.